src/goto-programs/goto_program_template.h

```diff
@@ -516,99 +477,60 @@ void goto_program_templatet<codeT, guardT>::compute_loop_numbers()
 }
 
 template <class codeT, class guardT>
-void goto_program_templatet<codeT, guardT>::get_successors(
-  targett target,
-  targetst &successors)
+template <typename Target>
+std::list<Target> goto_program_templatet<codeT, guardT>::get_successors(
+  Target target) const
 {
-  successors.clear();
   if(target==instructions.end())
-    return;
+    return std::list<Target>();
 
-  targett next=target;
-  next++;
+  const auto next=std::next(target);
 
   const instructiont &i=*target;
 
   if(i.is_goto())
   {
-    for(const auto &t : i.targets)
-      successors.push_back(t);
+    std::list<Target> successors(i.targets.begin(), i.targets.end());
 
     if(!i.guard.is_true() && next!=instructions.end())
       successors.push_back(next);
+
+    return successors;
   }
-  else if(i.is_start_thread())
+
+  if(i.is_start_thread())
   {
-    for(const auto &t : i.targets)
-      successors.push_back(t);
+    std::list<Target> successors(i.targets.begin(), i.targets.end());
 
     if(next!=instructions.end())
       successors.push_back(next);
+
+    return successors;
   }
-  else if(i.is_end_thread())
+
+  if(i.is_end_thread())
   {
     // no successors
+    return std::list<Target>();
   }
-  else if(i.is_throw())
+
+  if(i.is_throw())
   {
     // the successors are non-obvious
+    return std::list<Target>();
   }
-  else if(i.is_assume())
-  {
-    if(!i.guard.is_false() && next!=instructions.end())
-      successors.push_back(next);
-  }
-  else
-  {
-    if(next!=instructions.end())
-      successors.push_back(next);
-  }
-}
-
-template <class codeT, class guardT>
-void goto_program_templatet<codeT, guardT>::get_successors(
-  const_targett target,
-  const_targetst &successors) const
-{
-  successors.clear();
-  if(target==instructions.end())
-    return;
-
-  const_targett next=target;
-  next++;
-
-  const instructiont &i=*target;
 
-  if(i.is_goto())
+  if(i.is_assume() && !i.guard.is_false() && next!=instructions.end())
```

Comment:
You must not conflate `i.is_assume()` and the remainder of the test into one condition as this would yield a single-entry list either way.